Docs: How to interface external codes (#4093)
aiidateam/aiida-core: docs/source/howto/codes/interfacing.rst

.. _how-to:codes:plugin:

Interfacing external codes
==========================

In order to run an external code, you will first have to explain to AiiDA how to run it: how to create inputs files from the content of the input node, how the program can read those input files, where to find and how to read the output files, etc.

> The way you transmit these instructions to AiiDA is by extending the |CalcJob| and the |Parser| classes:

Reviewer: ```suggestion
The way you provide these instructions to AiiDA is by extending the |CalcJob| and the |Parser| classes:
```

.. code:: python

    # -*- coding: utf-8 -*-
    from aiida.engine import CalcJob

    class ArithmeticAddCalculation(CalcJob):

        @classmethod
        def define(cls, spec):
            super().define(spec)
            # Specifications of inputs, outputs and exit codes.
            # no return statement

        def prepare_for_submission(self, folder):
            # Instructions on how to prepare the input files
            return calcinfo

> 

Reviewer: You should place all code segments into context. Just a brief sentence before the segment about what you are showing.

Author: Not super sure what you mean since I think this code is preceeded by an explanation of what is going to contain, but I modified it a bit with the hope that it is clearer now.

.. code:: python

    # -*- coding: utf-8 -*-
    from aiida.parsers.parser import Parser

    class ArithmeticAddParser(Parser):

        def parse(self, **kwargs):
            # Instructions on how to process the output files
            # no return statement

> As you can guess from the name of the child-classes in the previous code snippet, we will illustrate the implementation of this code wrapping procedure with a possible design for the example introduced in the :ref:`tutorial<tutorial:basic>` of a code that sums two numbers.

Reviewer: I think it would be better to not directly reference the tutorial. For the how-to's, we make the assumption that people know the things we teach in the tutorial, but the how-to's are not a direct extension of the tutorial.

Think of it this way, an AiiDA expert might still want to look some things up among the how-to guides, but may never have gone through the tutorial.

Author: Ok; I think it still may be relevant if someone who went through the tutorial to place themselves more easily, but I reduced the presence to a "btw" comment that can be just ignored by more expert users:

> As you may have guessed from the name of the child-classes in the previous code snippet, we will illustrate this code wrapping procedure by designing a calculation that will add up two numbers (note that this is the same example used in the `tutorial`).

Would this be better or do you still think is best to remove it completely?

Reviewer: Yes, I think that's good.

> To have a clearer idea of how to interact with this code, we can assume it consists of the following bash script which basically reads two consecutive integers in a single line of an input file provided through a command line argument and write to standard output the result of its sum (so one would run it as ``$ add.sh -in [input_file] > [output_file]``).

Reviewer: I think you should introduce the example code with its purpose, inputs and outputs first. That will make it easier to follow how you wrap it.

> .. code:: bash

Reviewer: Consider placing the code for the `add.sh` script in  dropdown or possibly don't provide it at all. The how-to section is not a tutorial, so you can use examples, but there is no expectation that users can follow them step by step.

Reviewer 2: Here it might be better to go with the changes of Marnik, which does not require a separate script at all, but just uses `bash`. Since this can typically be found always at the same place, it makes it very easy to configure the code (and even provide a complete configuration yaml file) and easier to explain, as we are not executing some custom script, but just bash.

The next block seems similar but it is actually different: it is not defining new inputs but modifying some properties for inputs that are already defined in the base |CalcJob| class.
You can spot the difference in that defining is performed by using the ``spec.input`` method, singular, whereas accessing is achieved via the ``spec.inputs`` method, plural.
This change here is merely for convenience to provide `default` values for some relevant variables that will be needed for the next methods so that the user doesn't have to provide them when instantiating this |CalcJob| to run the code.
Note that these are not input nodes but options you set directly on the |CalcJob|, yet they are defined pretty much in the same way:

.. code:: python

        spec.input(
            'metadata.options.input_filename',
            valid_type=str,
            required=False,
            help='Filename to which the input for the code that is to be run is written.'
        )

You can see that this is input is defined with ``required=False``, so the user can ``run`` the code without specifying it in its instantiated |CalcJob|.

> This is not the case for the specific inputs we defined (since ``required=True`` by default), so if the user tries to ``run`` or ``submit`` the |CalcJob| without providing these, the code will raise.

Reviewer: ```suggestion
This is not the case for the specific inputs we defined (since ``required=True`` by default), so if the user tries to ``run`` or ``submit`` the |CalcJob| without providing these, the code will raise an exception.
```

> For more information on how to provide these different kind of inputs when runnig a code, you can see the :ref:`next how-to section<tutorial:basic>` which covers exactly this.

Reviewer: I'm confused about the reference here. Why are we linking to the tutorial?

Author: Sorry, I meant to reference the how-to run codes section, I copy/pasted and forgot to replace, he-he.

.. code:: python

    def prepare_for_submission(self, folder):

        input_x = self.inputs['x']
        input_y = self.inputs['y']
        input_code = self.inputs['code']
        input_filename = self.inputs['metadata']['options']['input_filename']
        output_filename = self.inputs['metadata']['options']['output_filename']

        with folder.open(input_filename, 'w', encoding='utf8') as handle:
            handle.write('{} {}\n'.format(input_x.value, input_y.value))

        codeinfo = CodeInfo()
        codeinfo.code_uuid = input_code.uuid
        codeinfo.stdout_name = output_filename
        codeinfo.cmdline_params = ['-in', input_filename]

        calcinfo = CalcInfo()
        calcinfo.codes_info = [codeinfo]
        calcinfo.retrieve_list = [output_filename]

        return calcinfo

The second block corresponds to our previously described first step (preparing the inputs to run): here we are using the ``folder.open`` method to get a handle to the file and write in an input file (in the format in which the code will read it) the values stored in our input nodes.

> This folder (along with all the files created in it) will not only be copied to the remote machine for the code to be ran there, but will also be stored in the local repository of the calculation node.

Reviewer: ```suggestion
This folder (along with all the files created in it) will not only be copied to the remote machine for the code to be run there, but will also be stored in the local repository of the calculation node.
```

The last two blocks correspond to the second step (instructions for the engine).
In the first one we are creating the ``codeinfo = CodeInfo()`` object, that will be then part of the ``calcinfo = CalcInfo()`` object, as one can see in ``calcinfo.codes_info = [codeinfo]``.
The ``code_uuid`` is necessary for the engine to get the base information set up in the |Code| node (in which machine the code is placed, what is the location of the executable, etc).
Note that this was taken from ``input_code = self.inputs['code']``, which we never specified in the ``define`` method: this is one of the two mandatory inputs defined in the base |CalcJob| that will need to be always provided in the instance.

> Moreover, in the case of the ``inputs['code']``, the designer of the |CalcJob| sub-class will always have to take care to pass its uuid into ``codeinfo.code_uuid`` as shown here.

Reviewer: ```suggestion
Moreover, in the case of the ``inputs['code']``, the designer of the |CalcJob| sub-class will always have to take care to pass its UUID into ``codeinfo.code_uuid`` as shown here.
```
The name UUID must be block capitalized.

Of the last 3 lines remaining, the first two are providing part of the instructions to the engine on how to build the running script:

.. code:: python

        codeinfo.stdout_name = output_filename
        codeinfo.cmdline_params = ['-in', input_filename]

The first one establishes the command line arguments to be passed to the code: in this case this is the only input of our ``add.sh`` script, but more complex codes can receive multiple inputs or flags through here.
The second one indicates to the engine where to redirect the standard output.
The combination of the two result in the following line in the running script:

.. code:: bash

    #!/bin/bash
    # Scheduler commands

    # Pre-running setups

    '${executable_from_code_uuid}' '-in' '${input_filename}' > ${output_filename}

    # Post-running commands

Which will default to ``'add.sh' '-in' 'aiida.in' > 'aiida.out'``.

> Through the |CodeInfo| instanse it is also posible to configure what to pass through the standard input, as well as what commands to run before and after the execution line.

Reviewer: ```suggestion
Through the |CodeInfo| instance it is also possible to configure what to pass through the standard input, as well as what commands to run before and after the execution line.
```

> The final line remaining (``calcinfo.retrieve_list = [output_filename]``) just indicates to the engine which files should be retrieved (either for immediate parsing or local storage for future post-processing).

Reviewer: Sphinx allows you to turn on line numbers for code segments. I'd suggest you turn those on an then reference the lines by number.

Author: Noise.
